File: tensorflow/core/util/tensor_bundle/byte_swap_tensor.cc
/* Copyright 2019 The TensorFlow Authors. All Rights Reserved.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
==============================================================================*/

#include "tensorflow/core/util/tensor_bundle/byte_swap_tensor.h"

#include <cstddef>
#include <cstring>
#include <string>

#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "tensorflow/core/framework/attr_value.pb.h"
#include "tensorflow/core/framework/function.pb.h"
#include "tensorflow/core/framework/graph.pb.h"
#include "tensorflow/core/framework/tensor.h"
#include "tensorflow/core/framework/tensor.pb.h"
#include "tensorflow/core/framework/types.h"
#include "tensorflow/core/framework/types.pb.h"
#include "tensorflow/core/platform/status.h"
#include "tensorflow/core/protobuf/meta_graph.pb.h"
#include "tensorflow/core/util/tensor_bundle/byte_swap_array.h"
#include "tsl/platform/errors.h"
#include "tsl/platform/logging.h"  // IWYU pragma: keep
#include "tsl/platform/mem.h"

namespace tensorflow {

namespace {

// Byte-swap a buffer in place.
//
// Args:
//  buff: pointer to the buffer to be modified IN PLACE.
//  size: size of bytes in this buffer.
//  dtype: type of data in this buffer.
//  num_of_elem: number of data in this buffer, set to -1 if it
//               could not be obtained directly from tensor data.
//               If num_of_elem is -1, this function will calculate
//               the number of data based on size and dtype.
// Returns: OkStatus() on success, -1 otherwise
absl::Status ByteSwapBuffer(char* buff, size_t size, DataType dtype,
                            int num_of_elem) {
  int array_len = num_of_elem;
  size_t bytes_per_elem = 0;

  switch (dtype) {
    // Types that don't need byte-swapping
    case DT_STRING:
    case DT_QINT8:
    case DT_QUINT8:
    case DT_BOOL:
    case DT_UINT8:
    case DT_INT8:
      return absl::OkStatus();

    // 16-bit types
    case DT_BFLOAT16:
    case DT_HALF:
    case DT_QINT16:
    case DT_QUINT16:
    case DT_UINT16:
    case DT_INT16:
      bytes_per_elem = 2;
      array_len = (array_len == -1) ? size / bytes_per_elem : array_len;
      break;

    // 32-bit types
    case DT_FLOAT:
    case DT_INT32:
    case DT_QINT32:
    case DT_UINT32:
      bytes_per_elem = 4;
      array_len = (array_len == -1) ? size / bytes_per_elem : array_len;
      break;

    // 64-bit types
    case DT_INT64:
    case DT_DOUBLE:
    case DT_UINT64:
      bytes_per_elem = 8;
      array_len = (array_len == -1) ? size / bytes_per_elem : array_len;
      break;

    // Complex types need special handling
    case DT_COMPLEX64:
      bytes_per_elem = 4;
      array_len = (array_len == -1) ? size / bytes_per_elem : array_len;
      array_len *= 2;
      break;

    case DT_COMPLEX128:
      bytes_per_elem = 8;
      array_len = (array_len == -1) ? size / bytes_per_elem : array_len;
      array_len *= 2;
      break;

    // Types that ought to be supported in the future
    case DT_RESOURCE:
    case DT_VARIANT:
      return absl::UnimplementedError(absl::StrCat(
          "Byte-swapping not yet implemented for tensors with dtype ", dtype));

    // Byte-swapping shouldn't make sense for other dtypes.
    default:
      return absl::UnimplementedError(absl::StrCat(
          "Byte-swapping not supported for tensors with dtype ", dtype));
  }

  TF_RETURN_IF_ERROR(ByteSwapArray(buff, bytes_per_elem, array_len));
  return absl::OkStatus();
}

}  // namespace

bool IsByteSwappable(DataType dtype) {
  switch (dtype) {
    // 16-bit types
    case DT_BFLOAT16:
    case DT_HALF:
    case DT_QINT16:
    case DT_QUINT16:
    case DT_UINT16:
    case DT_INT16:

    // 32-bit types
    case DT_FLOAT:
    case DT_INT32:
    case DT_QINT32:
    case DT_UINT32:

    // 64-bit types
    case DT_INT64:
    case DT_DOUBLE:
    case DT_UINT64:

    // Complex types
    case DT_COMPLEX64:
    case DT_COMPLEX128:
      return true;

    default:
      return false;
  }
}

absl::Status ByteSwapTensor(Tensor* t) {
  char* buff = const_cast<char*>((t->tensor_data().data()));
  return ByteSwapBuffer(buff, t->tensor_data().size(), t->dtype(),
                        t->NumElements());
}

absl::Status ByteSwapTensorContentInNode(NodeDef& node) {
  if (node.op() == "Const") {
    auto node_iterator = node.mutable_attr()->find("value");
    if (node_iterator != node.mutable_attr()->end()) {
      AttrValue node_value = node_iterator->second;
      if (node_value.has_tensor()) {
        auto tsize = node_value.mutable_tensor()->tensor_content().size();
        auto p_type = node_value.mutable_tensor()->dtype();
        // Swap only when there is something in tensor_content field
        if (tsize != 0 && DataTypeCanUseMemcpy(p_type)) {
          Tensor parsed(p_type);
          DCHECK(parsed.FromProto(*node_value.mutable_tensor()));
          if (!parsed.tensor_data().empty()) {
            TF_RETURN_IF_ERROR(ByteSwapTensor(&parsed));
            (*node.mutable_attr())["value"]
                .mutable_tensor()
                ->set_tensor_content(std::string(
                    reinterpret_cast<const char*>(parsed.tensor_data().data()),
                    parsed.tensor_data().size()));
          } else {
            void* copy = tsl::port::Malloc(tsize);
            std::memcpy(
                copy,
                std::string(node_value.mutable_tensor()->tensor_content())
                    .data(),
                tsize);
            TF_RETURN_IF_ERROR(ByteSwapBuffer((char*)copy, tsize, p_type, -1));
            (*node.mutable_attr())["value"]
                .mutable_tensor()
                ->set_tensor_content(
                    std::string(reinterpret_cast<const char*>(copy), tsize));
            tsl::port::Free(copy);
          }
        }
      }
    }
  }
  return absl::OkStatus();
}

absl::Status ByteSwapTensorContentInMetaGraphDef(MetaGraphDef* meta_graph_def) {
  for (auto& function : *meta_graph_def->mutable_graph_def()
                             ->mutable_library()
                             ->mutable_function())
    for (auto& node : (*function.mutable_node_def()))
      TF_RETURN_IF_ERROR(ByteSwapTensorContentInNode(node));
  return absl::OkStatus();
}

absl::Status ByteSwapTensorContentInGraphDef(GraphDef* graph_def) {
  for (auto& node : *graph_def->mutable_node())
    TF_RETURN_IF_ERROR(ByteSwapTensorContentInNode(node));
  return absl::OkStatus();
}

}  // namespace tensorflow

File: tensorflow/python/framework/byte_swap_tensor.py
# Copyright 2023 The TensorFlow Authors. All Rights Reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================

"""Utilities for byte swapping the tensor content."""

from tensorflow.core.framework import graph_pb2
from tensorflow.core.protobuf import meta_graph_pb2
from tensorflow.python.framework import dtypes

# Based on tensor_bundle/byte_swap.cc
byte_swappable = [
    dtypes.float16,
    dtypes.float32,
    dtypes.float64,
    dtypes.bfloat16,
    dtypes.complex64,
    dtypes.complex128,
    dtypes.uint16,
    dtypes.uint32,
    dtypes.uint64,
    dtypes.int16,
    dtypes.int32,
    dtypes.int64,
    dtypes.qint16,
    dtypes.quint16,
    dtypes.qint32,
]


def byte_swap_tensor_content(tensor, from_endiness, to_endiness):
  """Byte swaps.

  Args:
    tensor: Target tensor to change endiness.
    from_endiness: The original endianness format. "big" or "little"
    to_endiness: The target endianness format. "big" or "little"
  """
  if tensor.dtype in byte_swappable:
    tshape = tensor.tensor_shape.dim
    tensor_bytes = tensor.tensor_content
    if tensor_bytes:
      tensor_size = 1
      for sz in tshape:
        if sz.size != 0:
          tensor_size *= sz.size
      chunksize = len(tensor_bytes) // tensor_size
      # Split tensor_data into chunks for byte swapping.
      to_swap = [
          tensor_bytes[i : i + chunksize]
          for i in range(0, len(tensor_bytes), chunksize)
      ]
      # Swap and replace tensor_content.
      tensor.tensor_content = b"".join(
          [
              int.from_bytes(byteswap, from_endiness).to_bytes(
                  chunksize, to_endiness
              )
              for byteswap in to_swap
          ]
      )


def swap_tensor_content_in_graph_function(
    graph_def, from_endiness, to_endiness
):
  """Fix endiness of tensor contents.

  Args:
    graph_def: Target graph_def to change endiness.
    from_endiness: The original endianness format. "big" or "little"
    to_endiness: The target endianness format. "big" or "little"
  """
  if isinstance(graph_def, meta_graph_pb2.MetaGraphDef):
    functions = graph_def.graph_def.library.function
  elif isinstance(graph_def, graph_pb2.GraphDef):
    functions = graph_def.library.function
  else:
    return
  for function in functions:
    node_def = function.node_def
    for node in node_def:
      if node.op == "Const":
        tensor = node.attr["value"].tensor
        byte_swap_tensor_content(tensor, from_endiness, to_endiness)


def swap_tensor_content_in_graph_node(graph_def, from_endiness, to_endiness):
  for node in graph_def.node:
    if node.op == "Const":
      tensor = node.attr["value"].tensor
      byte_swap_tensor_content(tensor, from_endiness, to_endiness)
